### PR TITLE
Added partial collect tests

### PR DESCRIPTION
**Pull request by SidAnandMSFT**
## Issue

This is the issue tracking this: https://github.com/microsoft/Power-Fx/issues/1543

## Change description:
This PR adds a partial collect tests using only PowerFx recalEngine. The following is an explanation of the tests added by the this PR:
1. Powerfx_Collect_Without_Serielize
		a. Defines NewTable FormulaValue for a collection with a defined RecordType as schema type and list of NamedValues as value
		b. Defines recalEngine and symbol tables and runs `Last(collection).Col1` on it
		c. Used recalEngine to get the FormulaValue of the collection
		d. Define recalEngine and symbol table and runs `Collect` with a subset of the column keys of the collection
		e. Used the same recalEngine as above to run `Last(collection).Col1` and verifies the correct result
2. Powerfx_Collect_With_Serielize
	a. Defines NewTable FormulaValue for a collection with a defined RecordType as schema type and list of NamedValues as value
		b. Defines recalEngine and symbol tables and runs `Last(collection).Col1` on it
		c. Used recalEngine to get the FormulaValue of the collection (FormulaValue type = `RecordsOnlyTableValue`)
		d. Serielizes `RecordsOnlyTableValue` 
		e. Define a recalEngine to de-serielize the `RecordsOnlyTableValue`. We notice the output of de-serielization has the FormulaValue of type `InMemoryTableValue`
		f. Use the above recalEngine to run `Collect` with a subset of the column keys of the collection
		g. Used the same recalEngine as above to run `Last(collection).Col1`. This leads to an error of the type `Runtime type mismatch`

### CODE REVIEW

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs [hunk 1]
```diff
@@ -424,6 +424,84 @@ private void RecalcEngine_Symbol_CultureInfo6_ThreadProc(CultureInfo culture)
...
+        }
+
+        [Fact]
+        public void Powerfx_Collect_With_Serielize()
+        {
+            var result = (ErrorValue)Run_Collect_Worflow(true);
+            Assert.Equal("Runtime type mismatch", result.Errors[0].Message);
+        }
+
+        private FormulaValue Run_Collect_Worflow(bool serielize)
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs:441]: >serielize [](http://example.com/codeflow?start=54&length=9)

[typo] serielize ==> serialize

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs [hunk 2]
```diff
@@ -424,6 +424,84 @@ private void RecalcEngine_Symbol_CultureInfo6_ThreadProc(CultureInfo culture)
             Assert.Equal("Hello World!", helloWorld);
         }
 
+        [Fact]
+        public void Powerfx_Collect_Without_Serielize()
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs:428]: >Serielize [](http://example.com/codeflow?start=44&length=9)

[typo] Serielize ==> Serialize

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs [hunk 3]
```diff
@@ -424,6 +424,84 @@ private void RecalcEngine_Symbol_CultureInfo6_ThreadProc(CultureInfo culture)
...
+        {
+            var result = (BlankValue)Run_Collect_Worflow(false);
+            Assert.Equal("Blank()", result.ToString());
+        }
+
+        [Fact]
+        public void Powerfx_Collect_With_Serielize()
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs:435]: >Serielize [](http://example.com/codeflow?start=41&length=9)

[typo] Serielize ==> Serialize

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs [hunk 4]
```diff
@@ -424,6 +424,84 @@ private void RecalcEngine_Symbol_CultureInfo6_ThreadProc(CultureInfo culture)
...
+        {
+            var result = (ErrorValue)Run_Collect_Worflow(true);
+            Assert.Equal("Runtime type mismatch", result.Errors[0].Message);
+        }
+
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs:440]: Suggestion: merge those two [Fact]s into a single [Theory], and pass the differences as arguments to the method

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs [hunk 5]
```diff
@@ -424,6 +424,84 @@ private void RecalcEngine_Symbol_CultureInfo6_ThreadProc(CultureInfo culture)
...
+                AllowsSideEffects = true,
+                NumberIsFloat = true
+            };
+
+            // Define symbol table
+            var symbolTable = new SymbolTable();
+            symbolTable.EnableMutationFunctions();
+            var symbolValues = symbolTable.CreateValues();
+
+            if (!engineInit)
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs:493]: >engineInit [](http://example.com/codeflow?start=17&length=10)

Please rename it to something that is clearer. 'engineInit' can mean "initializeEngine" or "engineAlreadyInitialized", which are exact opposites. From the code it seems to be the latter, but we shouldn't have to look at the code to understand what a parameter means.

Another option would be to call it "setVariableValue", which is even clearer.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs [hunk 6]
```diff
@@ -424,6 +424,84 @@ private void RecalcEngine_Symbol_CultureInfo6_ThreadProc(CultureInfo culture)
...
+            formulaValue = recalEngine.GetValue("collection");
+
+            recalEngine = new RecalcEngine();
+            if (serielize)
+            {
+                // Serielizing it as SetPowerCardMemory does and stores it in memoryContext
+                string serielizedFormulaValue = formulaValue.ToExpression();
+
+                // Deserielizing it
+                // After serielization and deserielization, the formulaValue has type of InMemoryTableValue
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs:471]: >// After serielization and deserielization, the formulaValue has type of InMemoryTableValue [](http://example.com/codeflow?start=16&length=91)

This is an implementation detail, it should not have any effect on the test

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs [hunk 7]
```diff
@@ -424,6 +424,84 @@ private void RecalcEngine_Symbol_CultureInfo6_ThreadProc(CultureInfo culture)
...
+            var parserOptions = new ParserOptions()
+            {
+                AllowsSideEffects = true,
+                NumberIsFloat = true
+            };
+
+            // Define symbol table
+            var symbolTable = new SymbolTable();
+            symbolTable.EnableMutationFunctions();
+            var symbolValues = symbolTable.CreateValues();
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs:491]: >var symbolValues = symbolTable.CreateValues(); [](http://example.com/codeflow?start=12&length=46)

Unused

**Review by CarlosFigueiraMSFT — DISMISSED** [review]
:clock1:

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs [hunk 8]
```diff
@@ -424,6 +424,84 @@ private void RecalcEngine_Symbol_CultureInfo6_ThreadProc(CultureInfo culture)
...
+        {
+            var result = (BlankValue)Run_Collect_Worflow(false);
+            Assert.Equal("Blank()", result.ToString());
+        }
+
+        [Fact]
+        public void Powerfx_Collect_With_Serielize()
+        {
+            var result = (ErrorValue)Run_Collect_Worflow(true);
+            Assert.Equal("Runtime type mismatch", result.Errors[0].Message);
```
[the remark this answers is not in the document]
Comment (SidAnandMSFT) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs]: I added an issue to track this issue and added a link to the PR description

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/TableTests.cs [hunk 9]
```diff
@@ -24,5 +26,84 @@ public async Task TestTableWithEmptyRecords()
...
+        public void Powerfx_Partial_Collect(bool serialize)
+        {
+            FormulaValue result = Run_Collect_Worflow(serialize);
+            if (!serialize)
+            {
+                Assert.Equal("Blank()", result.ToString());
+            }
+            else
+            {
+                Assert.Equal("Runtime type mismatch", ((ErrorValue)result).Errors[0].Message);
```
Comment (anderson-joyle) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/TableTests.cs:42]: >Assert.Equal("Runtime type mismatch", ((ErrorValue)result).Errors[0].Message); [](http://example.com/codeflow?start=16&length=78)

Please remove this aseert. After the fixes both cases should result in `Blank()`

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/TableTests.cs [hunk 10]
```diff
@@ -24,5 +26,77 @@ public async Task TestTableWithEmptyRecords()
...
+            Assert.Equal("Col2", result.Value);
+
+            // Getting the FormulaValue that is used to update variable for the next expression
+            // formulaValue is of type RecordsOnlyTableValuee
+            formulaValue = recalEngine.GetValue("collection");
+
+            recalEngine = new RecalcEngine();
+            if (serialize)
+            {
+                // Serielizing it as SetPowerCardMemory does and stores it in memoryContext
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/TableTests.cs:65]: [typo] Serializing. Same in lines 69 and 70

**Review by anderson-joyle — DISMISSED** [review]
:shipit:

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/TableTests.cs [hunk 11]
```diff
@@ -24,5 +26,77 @@ public async Task TestTableWithEmptyRecords()
...
+            {
+                AllowsSideEffects = true,
+                NumberIsFloat = true
+            };
+
+            // Define symbol table
+            var symbolTable = new SymbolTable();
+            symbolTable.EnableMutationFunctions();
+
+            if (!setVariableValueInEngine)
```
[the remark this answers is not in the document]
Comment (SidAnandMSFT) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/TableTests.cs]: Agreed, swapped the way we use the flag

**Pull request merged**